frame/ibc/ibc-primitives/src/runtime_interface.rs

```diff
@@ -12,68 +15,149 @@ pub enum SS58CodecError {
-		use sp_trie::LayoutV0;
-		sp_trie::verify_trie_proof::<LayoutV0<sp_core::Blake2Hasher>, _, _, &[u8]>(
-			root,
-			proof,
-			&[(key, None)],
-		)
-		.is_ok()
-	}
+pub fn blake2_256_verify_non_membership_proof(root: &H256, proof: &[Vec<u8>], key: &[u8]) -> bool {
+	sp_trie::verify_trie_proof::<sp_trie::LayoutV0<sp_runtime::traits::BlakeTwo256>, _, _, &[u8]>(
+		root,
+		proof,
+		&[(key, None)],
+	)
+	.is_ok()
+}
 
-	fn sha2_512(message: &[u8]) -> [u8; 64] {
-		use sha2::Digest;
-		let mut hasher = sha2::Sha512::new();
-		hasher.update(message);
-		let hash = hasher.finalize();
-		let mut res = [0u8; 64];
-		res.copy_from_slice(&hash);
-		res
-	}
+pub fn sha2_512(message: &[u8]) -> [u8; 64] {
+	use sha2::Digest;
+	let mut hasher = sha2::Sha512::new();
+	hasher.update(message);
+	let hash = hasher.finalize();
+	let mut res = [0u8; 64];
+	res.copy_from_slice(&hash);
+	res
+}
 
-	fn sha2_512_truncated(message: &[u8]) -> [u8; 32] {
-		use sha2::Digest;
-		let mut hasher = sha2::Sha512::new();
-		hasher.update(message);
-		let hash = hasher.finalize();
-		let mut res = [0u8; 32];
-		res.copy_from_slice(&hash[..32]);
-		res
-	}
+pub fn sha2_512_truncated(message: &[u8]) -> [u8; 32] {
+	use sha2::Digest;
+	let mut hasher = sha2::Sha512::new();
+	hasher.update(message);
+	let hash = hasher.finalize();
+	let mut res = [0u8; 32];
+	res.copy_from_slice(&hash[..32]);
+	res
+}
 
-	fn sha3_512(message: &[u8]) -> [u8; 64] {
-		use sha3::Digest;
-		let mut hasher = sha3::Sha3_512::new();
-		hasher.update(message);
-		let hash = hasher.finalize();
-		let mut res = [0u8; 64];
-		res.copy_from_slice(&hash);
-		res
-	}
+pub fn sha3_512(message: &[u8]) -> [u8; 64] {
+	use sha3::Digest;
+	let mut hasher = sha3::Sha3_512::new();
+	hasher.update(message);
+	let hash = hasher.finalize();
+	let mut res = [0u8; 64];
+	res.copy_from_slice(&hash);
+	res
+}
 
-	fn ripemd160(message: &[u8]) -> [u8; 20] {
-		use ripemd::Digest;
-		let mut hasher = ripemd::Ripemd160::new();
-		hasher.update(message);
-		let hash = hasher.finalize();
-		let mut res = [0u8; 20];
-		res.copy_from_slice(&hash);
-		res
-	}
+pub fn ripemd160(message: &[u8]) -> [u8; 20] {
+	use ripemd::Digest;
+	let mut hasher = ripemd::Ripemd160::new();
+	hasher.update(message);
+	let hash = hasher.finalize();
+	let mut res = [0u8; 20];
+	res.copy_from_slice(&hash);
+	res
+}
+
+pub fn ss58_to_account_id_32(raw_str: &str) -> Result<[u8; 32], SS58CodecError> {
+	from_ss58check_with_version(raw_str)
+		.map(|acc| acc.into())
+		.map_err(|_| SS58CodecError::InvalidString)
+}
+
+pub fn account_id_to_ss58(bytes: [u8; 32]) -> Result<Vec<u8>, SS58CodecError> {
+	let account_id = AccountId32::new(bytes);
+	let encoded = to_ss58check_with_version(&account_id, 49); // todo: this should be based on the runtime.
+	Ok(encoded.as_bytes().to_vec())
+}
+
+// lifted directly from sp-core
+fn to_ss58check_with_version(account_id: &AccountId32, version: u16) -> String {
```

Comment:
shouldn't we have a test for this?

Reply:
or because it's a verbatim copy from sp-core, we don't need?

Reply:
good catch, will copy the tests